### PR TITLE
editors/kate: Kate Editor, syntax highlighting for Jakt

### PR DESCRIPTION
**Pull request by vheidari**
Hey there,

I have developed a `Jakt language` highlighting plugin for `Kate Editor`. I believe that this plugin would be a valuable addition to the main project. Therefore, I would like to kindly request a pull request to contribute my plugin to the main repository.

I have thoroughly tested the plugin and ensured that it meets the coding standards and guidelines of the Kate Editor project. It provides accurate syntax highlighting and enhances the editing experience for `Jakt language` files.

Thanks,
VHeidair

### CODE REVIEW

File: editors/kate/Jakt.xml [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+        <keywords casesensitive="1" />
+        <folding indentationsensitive="0" />
+        <emptyLines>
+            <emptyLine regexpr="\s+" />
+            <emptyLine regexpr="\s*#.*" />
+        </emptyLines>
+        <spellchecking>
+            <encoding char="á" string="\'a" />
+            <encoding char="à" string="\`a" />
+        </spellchecking>
```
Comment (alimpfard) [inline editors/kate/Jakt.xml:218]: This seems unrelated?

File: editors/kate/Jakt.xml [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+            <itemData name="MultipleComment" defStyleNum="dsComment" />
+        </itemDatas>
+    </highlighting>
+
+    <!-- Start General -->
+    <general>
+        <!-- Comments -->
+        <comments>
+            <comment name="singleLine" start="//" />
+            <comment name="multiLine" start="/*" end="*/" region="CommentFolding" />
```
Comment (alimpfard) [inline editors/kate/Jakt.xml:207]: This comment form doesn't actually exist.

File: editors/kate/Jakt.xml [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+            <!-- End handling multiple-line string : jaktSingleQuot -->
+            <context attribute="String" lineEndContext="#stay" name="jaktSingleQuot">
+                <DetectChar attribute="String" context="#pop" char="'" />
+            </context>
+            
+            <!-- End handling multiple-line comment : Comment -->
+            <context name="Comment" attribute="Comment" lineEndContext="#stay">
+                <DetectSpaces/>
+                <StringDetect attribute="Comment" context="#pop" String="*/" endRegion="comment"/>
+            </context>
```
Comment (alimpfard) [inline editors/kate/Jakt.xml:171]: ditto.

File: editors/kate/Jakt.xml [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+                
+            </context>
+            
+            <!-- End handling multiple-line string : jaktDoubleQuot -->
+            <context attribute="String" lineEndContext="#stay" name="jaktDoubleQuot">
+                <DetectChar attribute="String" context="#pop" char="&quot;" />
+            </context>
+            
+            <!-- End handling multiple-line string : jaktSingleQuot -->
+            <context attribute="String" lineEndContext="#stay" name="jaktSingleQuot">
```
Comment (alimpfard) [inline editors/kate/Jakt.xml:163]: This can optionally be preceded by `c` or `b`, for c_char and u8; not sure if the syntax highlighter needs to care though.

File: editors/kate/Jakt.xml [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+                <RegExpr attribute="Comment" context="assign" String="\/\/.*" firstNonSpace="false" />
+                
+                <!-- Start handling multiple-line string : jaktDoubleQuot -->               
+                <DetectChar attribute="String" context="jaktDoubleQuot" char="&quot;" />
+                
+                <!-- Start handling multiple-line string : jaktSingleQuot -->               
+                <DetectChar attribute="String" context="jaktSingleQuot" char="'" />
+        
+                <!-- Start handling JaktFunction -->
+                <RegExpr attribute="JaktFunction" context="assign" String="\w+[a-z-A-Z1-9](?=\s*\()" firstNonSpace="false" />
```
Comment (alimpfard) [inline editors/kate/Jakt.xml:132]: This will break with explicit generic parameters, though none of the other highlighters can accurately detect this either, so not a merge blocker.

File: editors/kate/Jakt.xml [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+        </list>
+        
+        
+        <!-- Start Contexts -->
+        <contexts>
+            <!-- Normal text context -->
+            <context attribute="NormalText" lineEndContext="#pop" name="NormalText">
+                <DetectSpaces />
+                <!-- Start handling multiple-line comments : Comment -->
+                <StringDetect attribute="Comment" context="Comment" String="/*" beginRegion="comment"/>
```
Comment (alimpfard) [inline editors/kate/Jakt.xml:120]: Same comment re no multiline comments.

File: editors/kate/Jakt.xml (hunk 1 above)
Comment (vheidari) [reply editors/kate/Jakt.xml:218]: Thank you, Ali. Yes, this part is unrelated to Jakt highlighting, and I will fix it in the next request.

File: editors/kate/Jakt.xml (hunk 2 above)
Comment (vheidari) [reply editors/kate/Jakt.xml:207]: Yep, I will fix it in next request.

File: editors/kate/Jakt.xml (hunk 4 above)
Comment (vheidari) [reply editors/kate/Jakt.xml:163]: I will work on it.

File: editors/kate/Jakt.xml (hunk 5 above)
Comment (vheidari) [reply editors/kate/Jakt.xml:132]: I work on it to prevent a break on explicit generic parameters ;)

File: editors/kate/Jakt.xml (hunk 6 above)
Comment (vheidari) [reply editors/kate/Jakt.xml:120]: ;)

File: editors/kate/Jakt.xml (hunk 3 above)
Comment (vheidari) [reply editors/kate/Jakt.xml:171]: ;)

**Comment by vheidari** [comment]
> The syntax directory is apparently `$HOME/.local/share/org.kde.syntax-highlighting/syntax/` according to https://invent.kde.org/frameworks/syntax-highlighting#syntax-definition-files (I tried `~/.local/share/kate/syntax` locally, it did not work with kate 23.08.3).
> 
> Missing things I noticed:
> 
>     * `fn` is not highlighted
> 
>     * `restricted(...)` is highlighted as a function
> 
>     * `never` isn't highlighted as a type
> 
>     * `let` is highlighted as a function in `let (foo, bar) = (a, b)`
> 
>     * `reflect` isn't highlighted (`reflect Type`)
> 
>     * The `weak` type prefix (`weak ClassType`) isn't highlighted
> 
> 
> That's about it :)

Hey Ali, thank you for the reply.

Oh, it's not working? I tested it on `Kate` Version `19.12.3` and it worked perfectly.

Let me check it on the newest version of `Kate Editor`. Thank you for mentioning it.

As for these items in the list (`fn`, `restricted(...)`, `never`, `let`,`reflect`, `weak`), I will resolve these problems :)

**Comment by vheidari** [comment]
According to the last comment by Ali regarding the path `~/.local/share/kate/syntax`, I checked it and found that `Kate` couldn't read the syntax file. To resolve this issue and allow `Kate` to read the **"Jakt.xml"** file, I have updated three valid paths that users can use. These paths have been updated in the `README.md` file.

    - $HOME/.local/share/katepart5/syntax
    - $HOME/.local/share/org.kde.syntax-highlighting/syntax
    - /usr/share/katepart5/syntax

Thank you, Ali, for bringing this to my attention.

I have also attached four images that demonstrate the syntax highlighter file tested on both Linux and Windows.
linux-usr-share-katepart5-syntax
windows-kate-jakt
linux-local-share-katepart5-syntax
linux-local-share-org kde syntax highlighting

**Comment by alimpfard** [comment]
Awesome, thanks!
Please squash the fixup/update commits and we'll be good to merge.

**Comment by vheidari** [comment]
Thank you @alimpfard ,
🔥️ All latest commits were squashed to first commit
